packages/vite/src/node/config.ts

```diff
@@ -381,10 +385,18 @@ export async function resolveConfig(
       return DEFAULT_ASSETS_RE.test(file) || assetsFilter(file)
     },
     logger,
-    createResolver
+    createResolver,
+    optimizeDeps: {
+      ...config.optimizeDeps,
+      esbuildOptions: {
+        keepNames: config.optimizeDeps?.keepNames,
+        external: config.optimizeDeps?.exclude,
+        ...config.optimizeDeps?.esbuildOptions
+      }
```

Comment:
I went for defaulting to the top-level options, but letting users replace them. Merging exclude with external could be surprising for users, if you set `external`, it should be respected.

Reply:
I think it may be better to include `external` in the list of excluded options. The rationale is that we may use `optimizeDeps.exclude` for something else (maybe logging, or some future feature). Since you can already achieve with `exclude` what you get with `external`, looks to me that this is the safer choice.

Reply:
Yeah, makes sense. Removed it.